src/racer/matchers.rs

```diff
@@ -658,11 +674,38 @@ fn find_doc(msrc: &str, blobend: usize) -> String {
         .collect::<Vec<_>>()  // These are needed because
         .iter()               // you cannot `rev`an `iter` that
         .rev()                // has already been `rev`ed.
-        .map(|line| if line.len() >= 4 {  // Remove "/// "
-                String::from(line[4..].to_owned())
+        .map(|line| {
+            let l = line.trim();
+            if l.len() >= 4 {  // Remove "/// "
+                String::from(l[4..].to_owned())
             } else {
                 String::new()
-            })
+            }
+        })
+        .collect::<Vec<_>>()
+        .join("\n")
+}
+
+fn find_mod_doc(msrc: &str, blobstart: usize) -> String {
+    let blob = &msrc[blobstart..];
+
+    blob.lines()
+        .take_while(|line| {
+            let l = line.trim();
+            l.starts_with("//") || l.len() == 0
+        })
+        // Skip over the copyright notice and empty lines until you find
+        // the module's documentation (it will go until the end of the
+        // file if the module doesn't have any docs).
+        .filter(|line| line.trim().starts_with("//! "))
+        .map(|line| {
+            let l = line.trim();
+            if l.len() >= 4 {  // Remove "/// "
+                String::from(l[4..].to_owned())
+            } else {
+                String::new()
+            }
+        })
         .collect::<Vec<_>>()
         .join("\n")
```

Comment:
There's several allocations happening in here including 2 from `String::from(l[4..].to_owned())` (the `String::from` and `to_owned` are redundant), and the `collect::<Vec<_>>()`. It may be more efficient to save this iterator to a temporary and then build a string like

```rust
let mut doc = String::new();
for line in doc_lines {
    doc.push_str(line);
    doc.push_str("\n");
}
```